### DISCUSSION TITLE
Should development continue?

### DISCUSSION THREAD

**Discussion by pops64** [pops64/Courseplay_FS22#10] [Polls]
If there are enough interested parties. I will continue development. Just understand this is not official CP and just a fork. But I am a former FS 17 CP developer so you are in good hands. I will do my best to keep it updated with official CP. Features that I was going to add but now only if there is enough support for this fork as some of theses require major time investments which I won't do if I am the only playing this fork.

- Fall back to turn on field if pathfinder fails(For those cases when there isn't enough room)
- Better traffic management
- Bug fixes that are just quirks and not game breaking
- Possibly two unload drivers following behind
- Maybe more

**Comment by coatsy35** [comment]
Just a thought, perhaps you could pull together a release in a Zip file ready for people who don't know how to use github. I myself comment here often but not sure how to zip up a development branch to test.

**Reply by pops64** [reply]
Yeah that is the plan. Hopefully I will get to that today. I want to fix the Hud elements first before packaging it.

**Reply by pops64** [reply]
Done. Can be found [here](https://github.com/pops64/Courseplay_FS22/releases/download/7.3.0.2-chopper/CourseplayChopper_FS22.zip) I have renamed it so both can be installed but there currently is no safety check so make sure to disable Official CP when launching

**Reply by coatsy35** [reply]
Will check it out tomorrow evening, late here in UK now.

**Comment by pops64** [comment]
Just an update. For those thinking or have voted for official CP merge, In order for that to happen you would have to convince Tensuko to permit it. Unfortunately, he much rather have no chopper support than something that doesn't do what he wants even if it works great. So you will have to organize the community and show Tensuko what your thoughts are. I tried creating a poll in official CP to get this merged and he deleted it immediately.

Also I would like a little more interest for new features but enough has been shown that I will at least finish the sharp turn logic and squash the bugs. This would be a state I consider as ready to merge. In the future I may make a poll for the features I was thinking of adding after it was merged into official CP and had more through testing. I will go by popularity of what should be done next.

**Comment by regs1980** [comment]
> Unfortunately, he much rather have no chopper support than something that doesn't do what he wants even if it works great. 

It works great, really?
I have shown them Part 1 of found errors.
I still found more than 20 bugs, which makes their fork a WIP alpha version.
They should have built on FS19 CP chopper support.
Reworked it to the point that it would fit the CP team for FS22. 

Stop shooting against the CP team, your fork is still in alpha version, present a chopper support worthy of coming into the CP.

**Reply by pops64** [reply]
The problem is Tensuko(not the CP team) just out right refuses to work with a work in progress and provide feedback. He also fails to understand that somethings are no longer practical either cause the code from FS19 didn't make it to 22. Specifically stuff needed to support turn maneuvers the way they use to work in 19 this would be a huge under taking to get ironed out not to mentioned they altered giants forage harvester base code in FS19 to get the code working the way it did.(I refuse to do this in all but the most extreme situations and usually with consent from all parties involved and I see no reason from a coding prescriptive to increase the range of the forage harvester to 40m) 

If you have any bugs or suggestions you would like in chopper support please create an issue. Compared to how other things currently work in CP I would say I am currently at the same quality or above(My version also fixes a lot of bugs in OG CP that aren't posted). Yes granted there are some glaring bugs but those would have been fixed prior to release. Specifically https://github.com/pops64/Courseplay_FS22/issues/15, https://github.com/pops64/Courseplay_FS22/issues/9, https://github.com/pops64/Courseplay_FS22/issues/3 and maybe https://github.com/pops64/Courseplay_FS22/issues/7 this one would take some doing as I mentioned earlier. All the other issues I created don't effect gameplay they would just be cool to have. Those always come last after a wider release to make sure you don't have any major bugs you didn't come across in testing that require a major shift in code logic. 

At the end of the day. I won't work with someone that instead of having a back and forth on what can and can not be accomplished just outright locks and closes the issue/PR and demands to do things that are either coding impossible or will no longer work in CP FS22. This is why I made my earlier comment. Especially after I have put a month of work into this.

**Comment by coatsy35** [comment]
Unfortunately I agree with @regs1980 get it in order with no bugs or no known bugs and code issues and then request a merge. I can see both points of view, but if this feature hits the quality benchmark then I can’t see why at that point they could refuse it. Chin up, you’ll get there. You’ve obviously got the skill and talent to do this, so I’m sure you can rise to the challenge

**Reply by pops64** [reply]
Thank you for the support. Like I said Tensuko refuses to listen and if I can't have access to the broader range of CP users cause chopper support is now an unofficial version instead of a PR, I will never get all the obvious bugs and I don't really want to add features that require major time investments for only 8 people. half of whom want to see it in CP

**Reply by coatsy35** [reply]
Don't give up. @Tensuku liked my post above. He's watching and so will many others. I think you can get it where it needs to be. If the bugs are that obvious I'm sure you'll get them all. And then even the CP team might help you eliminate the final ones.

**Reply by pops64** [reply]
It's an issue of how many edge cases there are in chopper operations. Like sharp turns I never tested them but I assumed my current logic handles them and it does in my test but I can see the issue with it. But there are problems with trying to fix it. The biggest issue is combines aren't very accurate with telling others when their turning just that they are or will be(This latter case is the problem when dealing with choppers). 

To make sharp turns better it would require making combines better or choppers better at sharing what stage of their turn they are in. Which would require figuring out how turns have changed since 17(Last time I took an indepth look at it) and figuring out a way to let others know what stage of turn they are in. Some of this may be in 19, If it is in 19 the logic might be right but chances are the functions/variables to support it are gone/changed/bugged from their original functionality. This was the problem with chase mode even after importing 19s code in attempt to fix it. A core function it relied on was bugged which was causing the code I made and the code from 19 to not work correctly.

Now you see why I need access to broader range of testers not just those interested in checking out an unofficial CP fork. Not only am I adding chopper support but I am changing upstream functions to either fix them or add functionality back which may or may not impact a broader range of things other than choppers. Like the chase mode upstream fix can effect every thing because it as to deal with where the front and rear of the vehicle, including attached implements, are which impacts turn maneuvers and whole bunch of other things. Test combines cause I did a lot of bug fixing with them.

You can also see why I don't really want to put the effort in because Tensuko thinks he as all the power and its not a team CP decision which is correct as he is the current maintainer but it was not how we did things back when Thomas was lead. Don't worry I still plan on fixing the bugs, sharp turns might take time but the others ones I have simple fixes for them. The fixes might not be the coolest to look at but it won't break gameplay

**Comment by coatsy35** [comment]
Combines must know where they are in a turn and broadcast it, as if you run two combines they following combine starts to slow down as the lead combine is making a turn, it then slows right down and waits until the lead combine has turned it then starts moving again. Try testing two combine path and see.

**Reply by pops64** [reply]
Yes and no. The follow combine code relies on distance apart. If the distance drops below this setting it will slow down/stop in an attempt to keep that distance.(There use to be a user setting for this somewhere) I think there is also a check for turn but this turn check is like I mentioned before only that the combine is turning or about to not what stage. Even my chopper unloader relies on this turn check. That is why the driver stops before the chopper at turns.

On up/down rows it doesn't cause issues. On sharp headland turns that check returns true at lot sooner more than likely causing you to lose fruit even if the animation is playing. The turn logic has no checks to stop the harvester if it can't discharge properly. So the way Giants code works is it still removes the fruit from game but then that fruit goes into digital oblivion as the rest of the code breaks down.

It comes down to figuring out where the chopper will be in an attempt to stay close enough and then managing the realignment after the turn. The best solution would be to disable the pocket that it makes so the turn is just a straight back and then turn. Thus eliminating the need to stay close as no fruit is processed. And then figuring out the alignment as there is safety checks that call if the unload driver is out of alignment with the chopper after a turn.

The another solution which I refuse to do, as there is no call for it, is to just make the range of the chopper discharge distance very large(This is the distance that trailers have of to be in range to be an eligible target). You then have to deal with ensuring this gets reset so human players can't use this to cheat. Not to mention the other issues you cause from this(particle effects gets weird, parked vehicles you don't want fruit in get fruit etc.)

**Comment by pops64** [comment]
Alright I am closing this thread out. Unfortunately, while there are some people that would like to see this continue. There just aren't enough total people interested. From here on out updates will come sporadically. Any game breaking bugs(Nil errors and the sort) will be fixed as soon as I get the message. 

If @Tensuko would like to rejoin to the conversation about merging this back into main CP I would love to hear from you. Please create an issue of anything you would like to see included that I haven't made already and comment on the ones that I do have. 

For now my next big push will grass support when I find the time to do it. Thanks all for those that have contributed to this discussion and to those that voted.
